### ISSUE TITLE
Overriding Systemd file on Debian Jessie breaks Docker

### ISSUE THREAD

**Issue by timwhite** [saltstack-formulas/docker-formula#80]
Debian jessie without specifying a version of docker gives us docker.io from backports which is 1.6.2 currently. This package provides a functioning systemd file.
```
$ cat /lib/systemd/system/docker.service
[Unit]
Description=Docker Application Container Engine
Documentation=http://docs.docker.com
After=network.target docker.socket
Requires=docker.socket

[Service]
EnvironmentFile=-/etc/default/docker
ExecStart=/usr/bin/docker -d -H fd:// $DOCKER_OPTS
MountFlags=slave
LimitNOFILE=1048576
LimitNPROC=1048576
LimitCORE=infinity

[Install]
WantedBy=multi-user.target
```

This formula however provides it's own systemd file which uses the ```docker daemon``` startup option that was introduced in docker 1.8, and so doesn't work with Jessie's 1.6
```
$ cat /etc/systemd/system/docker.service.d/docker-defaults.conf 
[Service]
EnvironmentFile=-/etc/default/docker
ExecStart=
ExecStart=/usr/bin/docker daemon $DOCKER_OPTS -H fd://
```

Obviously we can override the docker version to get a newer version not from backports that will work, but given that we provide the option for the backports version, we shouldn't override the systemd file.

Surely we can rely on the packaged systemd files instead of installing our own with this formula?

**Comment by blbradley** [comment]
I support using systemd unit files from packages.

This seems to have been added in #75 and broke support for Debian Jessie.

**Comment by blbradley** [comment]
Without specifying a version number, docker.io package version `1.6.2~dfsg1-1~bpo8+1` is installed (at this moment) which does not support `docker daemon`.

@tedski Can we remove `systemd.drop-in`? Or is there a good reason for keeping it?

**Comment by tedski** [comment]
The reason for `systemd.drop-in` is to support use of the `defaults` file.  The `defaults` file is where docker daemon runtime options are managed (see https://github.com/saltstack-formulas/docker-formula/blob/master/docker/init.sls#L91).

**Comment by blbradley** [comment]
The packages from Debian official and from Docker's repo already come with a systemd service unit file. Here is the one from Docker's repo for `1.12.2-0~jessie`:

`/etc/systemd/system/multi-user.target.wants/docker.service`:
```
[Unit]
Description=Docker Application Container Engine
Documentation=https://docs.docker.com
After=network.target docker.socket
Requires=docker.socket

[Service]
Type=notify
# the default is not to use systemd for cgroups because the delegate issues still
# exists and systemd currently does not support the cgroup feature set required
# for containers run by docker
ExecStart=/usr/bin/dockerd -H fd://
ExecReload=/bin/kill -s HUP $MAINPID
# Having non-zero Limit*s causes performance problems due to accounting overhead
# in the kernel. We recommend using cgroups to do container-local accounting.
LimitNOFILE=infinity
LimitNPROC=infinity
LimitCORE=infinity
# Uncomment TasksMax if your systemd version supports it.
# Only systemd 226 and above support this version.
#TasksMax=infinity
TimeoutStartSec=0
# set delegate yes so that systemd does not reset the cgroups of docker containers
Delegate=yes
# kill only the docker process, not all processes in the cgroup
KillMode=process

[Install]
WantedBy=multi-user.target
```

While this could be included in this repo with an `EnvironmentFile=-/etc/default/docker` statement to manage the runtime options, this would require tracking the changes of the unit file per version per package. It is probably more wise to let users manage this in their own states.

**Comment by tedski** [comment]
I agree that we shouldn't maintain an overriding unit file in the state.  We should, however, not break any current use of the `config` tree in pillar.  So, the question becomes, how to manage `config` and not maintain an entire unit file that overrides the package.  Is using `file.append` to add `EnvironmentFile` to the package's unit file just as bad?   Is there a more "systemd-way" to do this?

**Comment by blbradley** [comment]
Before any of my previous comments, I hadn't read about systemd drop-ins. My understanding is that users can put a `.conf` file into a folder for a particular service and override any systemd directive for that service. So, anyone can create an override in their own states. In addition, the syntax required for an EnvironmentFile is different from scripts in `/etc/default`. So, something like `export http_proxy="http://192.168.0.4:3128/"` (as shown in pillar example) would not set http_proxy. Thus, I strongly feel old sysvinit files and systemd EnvironmentFile need a strong separate.

**Comment by blbradley** [comment]
This may be a good time to mention `/etc/default/docker` in `1.12.2-0~jessie`:
```
# Docker Upstart and SysVinit configuration file

#
# THIS FILE DOES NOT APPLY TO SYSTEMD
#
#   Please see the documentation for "systemd drop-ins":
#   https://docs.docker.com/engine/articles/systemd/
#
```

**Comment by tedski** [comment]
I agree with the separation of sysvinit and systemd files, but both defaults files and EnvironmentFiles are simply sourced files.  So, I think it'd be best at this point to write the `config` dict to the appropriate file -- /etc/default/ in the case of sysvinit and a drop-in for systemd).

Also, the current example of `export http_proxy…` in `pillar.example` would work per the [docs](https://docs.docker.com/engine/admin/systemd/#/http-proxy) if it were capitalized properly, as it would be sourced as an environment variable from the defaults file.  Unless, of course, I'm misreading.

**Comment by blbradley** [comment]
Arguments to EnvironmentFile are not 'simply sourced'. When using EnvironmentFile, variable expansions do not take place, and shell expressions are not executed.

**Comment by blbradley** [comment]
I've started the work to fix and test support on Jessie. Please see #93.

**Comment by blbradley** [comment]
Closed in #93.
